polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -658,6 +659,47 @@ Value castToIndex(Value val) {
                                             builder.getIndexType(), val);
 }
 
+/// Unroll the given loop \p loop by \p factor.
+void unrollByFactor(LoopLikeOpInterface loop, unsigned factor) {
+  if (factor == 1)
+    return;
+
+  // Lowering affine for loop to scf for loop, as not all affine loops can be
+  // unrolled. Some affine loops (e.g., min expression upper bound) cannot be
+  // unrolled, as the lower bound of the cleanup loop cannot be expressed as an
+  // affine function.
+  if (Operation *loopOp = loop;
+      affine::AffineForOp affineForOp = dyn_cast<affine::AffineForOp>(loopOp)) {
+    OpBuilder builder(affineForOp);
+    Location loc = affineForOp.getLoc();
+    Value lowerBound = lowerAffineLowerBound(affineForOp, builder);
+    Value upperBound = lowerAffineUpperBound(affineForOp, builder);
+    Value step =
+        builder.create<arith::ConstantIndexOp>(loc, affineForOp.getStep());
+    auto scfForOp = builder.create<scf::ForOp>(
+        loc, lowerBound, upperBound, step, affineForOp.getIterOperands());
+
+    // Replace scfForOp body with affineForOp body.
+    scfForOp.getBody()->getParent()->getBlocks().remove(scfForOp.getBody());
+    scfForOp.getRegion().getBlocks().splice(
+        scfForOp.getRegion().end(), affineForOp.getRegion().getBlocks());
+
+    // Replace affine::AffineYieldOp with scf::YieldOp.
+    scfForOp.walk([&](affine::AffineYieldOp yieldOp) {
+      OpBuilder builder(yieldOp);
+      builder.create<scf::YieldOp>(yieldOp.getLoc(), yieldOp.getOperands());
+      yieldOp.erase();
+    });
+
+    affineForOp->replaceAllUsesWith(scfForOp);
+    affineForOp->erase();
+    loop = scfForOp;
+  }
+
+  LogicalResult res = loopUnrollByFactor(cast<scf::ForOp>(loop), factor);
+  assert(res.succeeded() && "Expecting tiled loop to be unrolled");
```

Comment:
```suggestion
  assert(succeeded(loopUnrollByFactor(cast<scf::ForOp>(loop), factor)) && "Expecting tiled loop to be unrolled");
```
NIT: I like this better, but it is really a super NIT.

Reply:
> I worry that unrolling is not perform when the compile is built with assert off.

... yes, and technically, you even want a `(void)res;` afterwards to prevent warnings about the unused variable when assertions are off.

Reply:
Added `[[maybe_unused]]`

Reply:
True, this is the way to go :+1: